conformance/utils/kubernetes/helpers.go

```diff
@@ -777,6 +777,68 @@ func TLSRouteMustHaveCondition(t *testing.T, client client.Client, timeoutConfig
 	require.NoErrorf(t, waitErr, "error waiting for TLSRoute status to have a Condition matching expectations")
 }
 
+// GatewayAndUDPRoutesMustBeAccepted waits until the specified Gateway has an IP
+// address assigned to it and the UDPRoute has a ParentRef referring to the
+// Gateway. The test will fail if these conditions are not met before the
+// timeouts.
+func GatewayAndUDPRoutesMustBeAccepted(t *testing.T, c client.Client, timeoutConfig config.TimeoutConfig, controllerName string, gw GatewayRef, routeNNs ...types.NamespacedName) string {
+	t.Helper()
+
+	gwAddr, err := WaitForGatewayAddress(t, c, timeoutConfig, gw.NamespacedName)
+	require.NoErrorf(t, err, "timed out waiting for Gateway address to be assigned")
+
+	ns := gatewayv1.Namespace(gw.Namespace)
+	kind := gatewayv1.Kind("Gateway")
+
+	for _, routeNN := range routeNNs {
+		namespaceRequired := true
+		if routeNN.Namespace == gw.Namespace {
+			namespaceRequired = false
+		}
+
+		var parents []gatewayv1.RouteParentStatus
+		for _, listener := range gw.listenerNames {
+			parents = append(parents, gatewayv1.RouteParentStatus{
+				ParentRef: gatewayv1.ParentReference{
+					Group:       (*gatewayv1.Group)(&gatewayv1.GroupVersion.Group),
+					Kind:        &kind,
+					Name:        gatewayv1.ObjectName(gw.Name),
+					Namespace:   &ns,
+					SectionName: listener,
+				},
+				ControllerName: gatewayv1.GatewayController(controllerName),
+				Conditions: []metav1.Condition{
+					{
+						Type:   string(gatewayv1.RouteConditionAccepted),
+						Status: metav1.ConditionTrue,
+						Reason: string(gatewayv1.RouteReasonAccepted),
+					},
+				},
+			})
+		}
+		UDPRouteMustHaveParents(t, c, timeoutConfig, routeNN, parents, namespaceRequired)
+	}
```

Comment:
This portion of the code is duplicated from `GatewayAndHTTPRoutesMustBeAccepted` and `GatewayAndTLSRoutesMustBeAccepted`. I think it'd be cool to extract this check on listeners and maybe pass `UDPRouteMustHaveParents`, `HTTPRouteMustHaveParents`, and `TLSRouteMustHaveParents` handlers as parameters to the shared function. I didn't dig too much into the code, but at first glance, it looks doable.

Reply:
Good idea, that seems doable to me as well, let's try that
The only difference is that `GatewayAndHTTPRoutesMustBeAccepted` checks for listener conditions, but that could be applied to TLSRoutes and UDPRoutes as well.